File: OOP/tic-top-toe-app/src/com/techclab/game/Game.java
package com.techclab.game;

import com.techlab.board.Board;
import com.techlab.enums.GameStatus;
import com.techlab.enums.Mark;
import com.techlab.exceptions.CellIsAlredyMarkedException;
import com.techlab.player.Player;
import com.techlab.resultAnalyzer.ResultAnalyzer;

public class Game {

	private Board board;
	private ResultAnalyzer analyzer;
	private Player p1, p2, currentPlayer;
	private GameStatus status = GameStatus.INPROGRESS;
	private int moveCounter = 1;

	public Game(Player one, Player two) {
		board = new Board();
		analyzer = new ResultAnalyzer(board);
		p1 = one;
		p2 = two;
		currentPlayer = p1;
	}

	public int getMOVE_COUNTER() {
		return moveCounter;
	}

	public void makeMove(int cellNumber) throws CellIsAlredyMarkedException {
		board.putMarkOnCell(currentPlayer.getMark(), cellNumber);
		moveCounter++;
	}

	public GameStatus play(int cellNumber) throws CellIsAlredyMarkedException {
		currentPlayer = getCurrentPlayer();
		makeMove(cellNumber);
		if (analyzer.checkForWin()) {
			status = GameStatus.WIN;
		}
		if (analyzer.checkForDraw()) {
			status = GameStatus.DRAW;
		}
		return status;
	}

	public Player getCurrentPlayer() {
		if (moveCounter % 2 == 0) {
			return p2;
		}
		return p1;
	}
	
	public Player getCurrentTurn() {
		return currentPlayer;
	}

	public void displayBoard() {
		for (int index = 0; index < board.getBoardSize(); index++) {
			System.out.println(board.getMarkOnCell(index++) + "\t" + board.getMarkOnCell(index++) + "\t"
					+ board.getMarkOnCell(index));
		}
	}

	public GameStatus getStatus() {
		return status;
	}

}

File: nodejs/tic-tac-toe/Game.js
let Board = require('./board');
let Mark = require('./Mark');
let Player = require('./player');
let GameStatus = require('./gameStatus');
let ResultAnalizer = require('./resultAnalyzer');

class Game {
    constructor(){
        this.board = new Board();
        this.resultAnalizer = new ResultAnalizer(this.board); 
        this.p1 = new Player('vivek',Mark.O);
        this.p2 = new Player('kishan',Mark.X);
        this.currentPlayer = this.p1;
        this.moveCounter = 0;
        this.gameStatus = GameStatus.InProgress;
    }

   

    makeMove(cellNumber){
        this.board.markOnCell(cellNumber, this.currentPlayer.mark);
        this.moveCounter++;
        this.checkGameStatus();
    }

    checkGameStatus(){
        if(this.isGameWin()){
            this.gameStatus = GameStatus.Win;
        }
        if(this.isGameDraw()){
            this.gameStatus = GameStatus.Draw;
        }
        if(this.gameStatus == GameStatus.InProgress){
            this.switchPlayer();
        }
    }

    isGameWin(){
        return this.resultAnalizer.checkForWin();
    }

    isGameDraw(){
        return this.resultAnalizer.checkForDraw();
    }

    displayBoard(){
        this.board.printBoard();
    }

    switchPlayer(){
        if((this.moveCounter % 2) == 0){
            this.currentPlayer = this.p1;
        }else{
            this.currentPlayer = this.p2;
        }
    }
}

module.exports = Game;